### PR TITLE
When restarting a partial join resync, prioritise the server which actioned a partial join

### LINKED ISSUE #12999: Faster joins: smarter algorithm for picking a server to resync from
Per https://github.com/matrix-org/synapse/pull/12813#discussion_r878091275: when we restart resyncing a room, it can take a while before we find an appropriate server.

https://github.com/matrix-org/synapse/blob/7c6b2204d143550d81e5bf9612c4e69fe0866b4c/synapse/handlers/federation.py#L1517-L1520

### PR DESCRIPTION
**Pull request by DMRobertson**
Fixes #12999. See https://github.com/matrix-org/synapse/issues/12999#issuecomment-1273325438 and the comments below it for context.

Should be commitwise reviewable.

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -1708,7 +1708,7 @@ async def _sync_partial_state_room(
                             destination, event
                         )
                         break
-                    except FederationError as e:
+                    except (NotRetryingDestination, FederationError) as e:
```
Comment (DMRobertson) [inline synapse/handlers/federation.py:1707]: I have no idea if there are other exceptions we ought to catch here.

File: synapse/storage/database.py [hunk 2]
```diff
@@ -1658,7 +1658,7 @@ async def simple_select_one_onecol(
             table: string giving the table name
             keyvalues: dict of column names and values to select the row with
             retcol: string giving the name of the column to return
-            allow_none: If true, return None instead of failing if the SELECT
+            allow_none: If true, return None instead of raising StoreError if the SELECT
```
Comment (DMRobertson) [inline synapse/storage/database.py:1661]: This was a drive-by.

File: synapse/storage/databases/main/room.py [hunk 3]
```diff
@@ -1182,7 +1200,12 @@ async def get_partial_state_rooms_and_servers(
         for row in rows:
             room_id = row["room_id"]
             server_name = row["server_name"]
-            room_servers.setdefault(room_id, []).append(server_name)
+            # There is a foreign key constraint which enforces that every room_id in
+            # partial_state_rooms_servers appears in partial_state_rooms, so this lookup
+            # should always succeed.
+            # TODO(faster joins): To make this robust we should run both SELECTs in the
+            #   same transaction.
```
Comment (DMRobertson) [inline synapse/storage/databases/main/room.py:1207]: I left this as a TODO for pragmatism's sake (read: I don't want to read database.py right now). I'm assuming it's very unlikely that a new partial join will have started and completed between the two SELECTs.

File: synapse/handlers/federation.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/federation.py:1707]: I think this is redundant.

`update_state_for_partial_state_event` calls `_compute_event_context_with_maybe_missing_prevs`, which catches all `Exceptions` and turns them into `FederationError`s. 

I don't think there's anywhere else in `update_state_for_partial_state_event` that can raise a `NotRetryingDestination`.

File: synapse/storage/schema/main/delta/73/09partial_joined_via_destination.sql [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- When we resync partial state, we prioritise doing so using the server we
+-- partial-joined from. To do this we need to record which server that was!
+-- Both columns are backwards compatible.
```
Comment (richvdh) [inline synapse/storage/schema/main/delta/73/09partial_joined_via_destination.sql:18]: both columns? which columns?

File: synapse/storage/databases/main/room.py (hunk 3 above)
Comment (richvdh) [reply synapse/storage/databases/main/room.py:1207]: Note that a transaction only helps you at a `REPEATABLE READ` (or higher) isolation level (which, in fairness, we do use by default, but every so often we talk about reducing it).

I'd just do `room_servers.get(room_id)` and add a check for `None`, to save having to worry about it.

File: synapse/handlers/federation.py [hunk 5]
```diff
@@ -1742,3 +1737,21 @@ async def _sync_partial_state_room(
...
+
+def _prioritise_destinations_for_partial_state_resync(
+    initial_destination: Optional[str], other_destinations: Collection[str]
+) -> Sequence[str]:
+    """Work out the order in which we should ask servers to resync events.
+
+    If an `initial_destination` is given, it takes top priority. Otherwise
+    all servers are treated equally.
+    """
+    if initial_destination is not None:
```
Comment (richvdh) [inline synapse/handlers/federation.py:1750]: suggest inverting this condition for simplicity:

```suggestion
    if initial_destination is None:
        return other_destinations
```

File: synapse/handlers/federation.py [hunk 6]
```diff
@@ -1617,7 +1617,7 @@ async def _resume_sync_partial_state_room(self) -> None:
     async def _sync_partial_state_room(
         self,
         initial_destination: Optional[str],
-        other_destinations: Collection[str],
+        other_destinations: Sequence[str],
```
Comment (richvdh) [inline synapse/handlers/federation.py:1620]: can you explain why this is needed? `_prioritise_destinations_for_partial_state_resync` takes a `Collection`, not a `Sequence`.

File: synapse/handlers/federation.py [hunk 7]
```diff
@@ -1651,15 +1652,9 @@ async def _sync_partial_state_room(
...
-        destinations: Collection[str]
-        if initial_destination is not None:
-            # Move `initial_destination` to the front of the list.
-            destinations = list(other_destinations)
-            if initial_destination in destinations:
-                destinations.remove(initial_destination)
-            destinations = [initial_destination] + destinations
-        else:
-            destinations = other_destinations
+        destinations = _prioritise_destinations_for_partial_state_resync(
```
Comment (richvdh) [inline synapse/handlers/federation.py:1655]: if we're factoring out this logic... wouldn't it make sense to move the constraint check just above (`if initial_destination is None and len(other_destinations) == 0:`) out too?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: synapse/storage/schema/main/delta/73/09partial_joined_via_destination.sql (hunk 4 above)
Comment (DMRobertson) [reply synapse/storage/schema/main/delta/73/09partial_joined_via_destination.sql:18]: Copy-paste leftover from the last schema I wrote. Will fix.

File: synapse/handlers/federation.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/handlers/federation.py]: In that case I'm confused about what you meant by

> So, I'd just ignore destinations for now, and as @MadLittleMods, just make sure we catch the relevant exceptions, and move on.

in https://github.com/matrix-org/synapse/issues/12999#issuecomment-1273325438. Perhaps the point is that your reasoning in the comment above proves that we're catching the relevant exceptions?

In any case, 29b362dad reverts this.

File: synapse/storage/databases/main/room.py (hunk 3 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/room.py:1207]: Thanks---the comment is naive as written.

(The reference here is https://www.postgresql.org/docs/current/transaction-iso.html .)

File: synapse/storage/schema/main/delta/73/09partial_joined_via_destination.sql (hunk 4 above)
Comment (DMRobertson) [reply synapse/storage/schema/main/delta/73/09partial_joined_via_destination.sql]: 468c5c2f6

File: synapse/handlers/federation.py (hunk 7 above)
Comment (DMRobertson) [reply synapse/handlers/federation.py:1655]: Sure.

FWIW: my motivation for pulling it out is that `_sync_partial_state_room` is a long[^1] function and this helps to keep it short. Looking back on the change though: it's pretty arguable. No strong opinions.

File: synapse/storage/databases/main/room.py (hunk 3 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/room.py]: Many thanks. This is 9e9be2b787

File: synapse/handlers/federation.py (hunk 6 above)
Comment (DMRobertson) [reply synapse/handlers/federation.py:1620]: I think this might have been a left-over from a WIP version before I pulled out the helper. (Or maybe I was trying to express "we preserve the iteration order of `other_destinations`"? :shrug: )

File: synapse/handlers/federation.py (hunk 5 above)
Comment (DMRobertson) [reply synapse/handlers/federation.py]: 1c9988ee36

File: synapse/handlers/federation.py (hunk 6 above)
Comment (DMRobertson) [reply synapse/handlers/federation.py:1620]: 1ccf9b8063, in any case

File: synapse/handlers/federation.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/federation.py]: Yes, it looks like we're *already* catching the relevant exceptions. I was just saying we should check that.

**Review by richvdh — APPROVED** [review]
lgtm
